bots/forward/src/main/java/org/openjdk/skara/bots/forward/ForwardBot.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.skara.bots.forward;
+
+import org.openjdk.skara.bot.*;
+import org.openjdk.skara.host.*;
+import org.openjdk.skara.vcs.*;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Files;
+import java.net.URLEncoder;
+import java.util.List;
+import java.util.logging.Logger;
+
+class ForwardBot implements Bot, WorkItem {
+    private final Logger log = Logger.getLogger("org.openjdk.skara.bots");;
+
+    private final HostedRepository fromHostedRepo;
+    private final Branch fromBranch;
+
+    private final HostedRepository toHostedRepo;
+    private final Branch toBranch;
+
+    ForwardBot(HostedRepository fromHostedRepo, Branch fromBranch,
+               HostedRepository toHostedRepo, Branch toBranch) {
+        this.fromHostedRepo = fromHostedRepo;
+        this.fromBranch = fromBranch;
+        this.toHostedRepo = toHostedRepo;
+        this.toBranch = toBranch;
+    }
+
+    @Override
+    public boolean concurrentWith(WorkItem other) {
+        if (!(other instanceof ForwardBot)) {
+            return true;
+        }
+        var otherBot = (ForwardBot) other;
+        return !toHostedRepo.getName().equals(otherBot.toHostedRepo.getName());
+    }
+
+    @Override
+    public void run(Path scratchPath) {
+        try {
+            var sanitizedUrl =
+                URLEncoder.encode(toHostedRepo.getUrl().toString(), StandardCharsets.UTF_8);
+            var toDir = scratchPath.resolve("forward").resolve(sanitizedUrl);
```

Comment:
Perhaps it would be better to use permanent storage here (like the hgbridge) to avoid excessive copies when running concurrently.

Reply:
Good idea, I pushed a new commit that utilizes shared storage instead.